OADP-4879 Added information for BSL config (#82961)
openshift/openshift-docs: modules/oadp-configuring-aws-bsl.adoc

// Module included in the following assemblies:
//
// * backup_and_restore/application_backup_and_restore/installing/installing-oadp-aws.adoc

:_mod-docs-content-type: PROCEDURE

> [id="oadp-configuring-aws-bsl.adoc_{context}"]

Reviewer: The ID shouldn't have .adoc in it:
```suggestion
[id="oadp-configuring-aws-bsl_{context}"]
```

= Configuring the backup storage location using AWS

> You can configure the {aws-short} backup storage location (BSL) as shown in the following example.

Reviewer: Nit - not sure this is a big deal but maybe saying example procedure would be clearer:
```suggestion
You can configure the {aws-short} backup storage location (BSL) as shown in the following example procedure.
```

Up to you though

.Prerequisites

* You have created an object storage bucket using {aws-short}.
* You have installed the {oadp-short} Operator.

.Procedure

> . Configure the BSL custom resource as shown in the following example:

Reviewer 2: ```suggestion
. Configure the BSL custom resource (CR) as shown in the following example:
```

> . Configure the BSL custom resource (CR) as shown in the following example:

Reviewer: Since this is the only step in the procedure, this should be an unnumbered list item [per the SSG](https://redhat-documentation.github.io/supplementary-style-guide/#single-step-procedures):
```suggestion
* Configure the BSL custom resource (CR) as shown in the following example:
```

+

> .Example backup storage location

Reviewer: This generally looks good, but I think one issue is that there are multiple references to this being an example, while this yaml file has user-replaced values instead of actual example values. I would either ask an SME for example values, or change the wording in this line and in the step to not refer to this as an example

Author: re-worded it. thanks.

> [source,yaml]

Reviewer 3: We do not config the BSL resource directly
instead we configure the BSL in the DataProtetionApplication rescoues
@shdeshpa07

Author: Thanks Amos. This BSL is a description of all the fields that can be configured. You are right, we generally configure the BSL in the DPA. There are already examples of configuring the BSL in the DPA in our doc set. I have linked those examples as additional resources. Does this sound good to you Thanks.

----
apiVersion: oadp.openshift.io/v1alpha1
kind: BackupStorageLocation
metadata:
  name: default
  namespace: openshift-adp
spec:
  provider: aws # <1>
  objectStorage:
    bucket: <bucket_name> # <2>
    prefix: <bucket_prefix> # <3>
  credential: # <4>
    key: cloud # <5>
    name: cloud-credentials # <6>
  config:
    region: <bucket_region> # <7>
    s3ForcePathStyle: "true" # <8>
    s3Url: <s3_url> # <9>
    publicUrl: <public_s3_url> # <10>
    serverSideEncryption: AES256 # <11>
    kmsKeyId: "50..c-4da1-419f-a16e-ei...49f" # <12>    
    customerKeyEncryptionFile: "/credentials/customer-key" # <13>
    signatureVersion: "1" # <14>
    profile: "default" # <15>
    insecureSkipTLSVerify: "true" # <16>
    enableSharedConfig: "true" # <17>
    tagging: "" # <18>
    checksumAlgorithm: "CRC32" # <19>
----

> <1> The name of the object store plugin. In this example, the plugin in `aws`. This field is required.

Reviewer 2: ```suggestion
<1> The name of the object store plugin. In this example, the plugin is `aws`. This field is required.
```

<2> The name of the bucket in which to store backups. This field is required.
<3> The prefix within the bucket in which to store backups. This field is optional.

> <4> The credentials for the backup storage location. You can set custom credentials. If custom credentials are not set, the default credentials secret is used.

Reviewer 2: ```suggestion
<4> The credentials for the backup storage location. You can set custom credentials. If custom credentials are not set, the default credential's secret is used.
```
OR
```suggestion
<4> The credentials for the backup storage location. You can set custom credentials. If custom credentials are not set, the default credentials' secret is used.
```

> <5> The `key` within the secret credentials data.

Reviewer 2: ```suggestion
<5> The `key` within the secret credential's data.
```
OR
```suggestion
<5> The `key` within the secret credentials' data.
```

<6> The name of the secret containing the credentials.
<7> The AWS region where the bucket is located. Optional if s3ForcePathStyle is false.

> <8> A boolean flag to decide whether to use path-style addressing instead of virtual hosted bucket addressing. Set to "true" if using a storage service such as MinIO, or Noobaa. This is an optional field. The default value is `false`.

Reviewer 2: ```suggestion
<8> A boolean flag to decide whether to use path-style addressing instead of virtual hosted bucket addressing. Set to "true" if using a storage service such as MinIO or Noobaa. This is an optional field. The default value is `false`.
```

Reviewer 2: Also, please can you check the capitalization of Noobaa - see https://www.noobaa.io/

Author: Yes, you are correct. It is NooBaa. I'll make that change everywhere.

> <8> A boolean flag to decide whether to use path-style addressing instead of virtual hosted bucket addressing. Set to "true" if using a storage service such as MinIO or NooBaa. This is an optional field. The default value is `false`.

Reviewer: So this is a little tricky, but even if the value is "true" with quotes in the YAML, it still needs to be in backticks because it's part of a file:
```suggestion
<8> A boolean flag to decide whether to use path-style addressing instead of virtual hosted bucket addressing. Set to `"true"` if using a storage service such as MinIO or NooBaa. This is an optional field. The default value is `false`.
```

Reviewer: Or if this was simply meant to be backticks without quotes, you can do that as well.

> <9> You can specify the {aws-short} S3 URL here for explicitness. This field is primarily for storage services such as MinIO, or Noobaa. This is an optional field.

Reviewer 2: ```suggestion
<9> You can specify the {aws-short} S3 URL here for explicitness. This field is primarily for storage services such as MinIO or Noobaa. This is an optional field.
```

> <10> This field is primarily used for storage services such as MinIO, or Noobaa. This is an optional field.

Reviewer 2: ```suggestion
<10> This field is primarily used for storage services such as MinIO or Noobaa. This is an optional field.
```

<11> The name of the server-side encryption algorithm to use for uploading objects, for example, `AES256`. This is an optional field.

> <12> Specify an {aws-short} KMS key ID. You can format as shown in the example, as an alias, such as `alias/<KMS-key-alias-name>`, or the full `ARN` to enable encryption of the backups stored in S3. Note that, `kmsKeyId` cannot be used in with `customerKeyEncryptionFile`. This is an optional field.

Reviewer 2: ```suggestion
<12> Specify an {aws-short} KMS key ID. You can format, as shown in the example, as an alias, such as `alias/<KMS-key-alias-name>`, or the full `ARN` to enable encryption of the backups stored in S3. Note that, `kmsKeyId` cannot be used in with `customerKeyEncryptionFile`. This is an optional field.
```

> <12> Specify an {aws-short} KMS key ID. You can format, as shown in the example, as an alias, such as `alias/<KMS-key-alias-name>`, or the full `ARN` to enable encryption of the backups stored in S3. Note that, `kmsKeyId` cannot be used in with `customerKeyEncryptionFile`. This is an optional field.

Reviewer: I think this comma could be removed here:
```suggestion
<12> Specify an {aws-short} KMS key ID. You can format, as shown in the example, as an alias, such as `alias/<KMS-key-alias-name>`, or the full `ARN` to enable encryption of the backups stored in S3. Note that `kmsKeyId` cannot be used in with `customerKeyEncryptionFile`. This is an optional field.
```

> <13> Specify the file that has the `SSE-C` customer key to enable customer key encryption of the backups stored in S3. The file must contain a 32-byte string. The `customerKeyEncryptionFile` points to a mounted secret within the `velero` container. Add the following key-value pair to the `velero` `cloud-credentials` secret: `customer-key: <your_b64_encoded_32byte_string>`. Note that the `customerKeyEncryptionFile` field cannot be used in with the `kmsKeyId`. The default value is an empty string (""), which means, `SSE-C` is disabled. This is an optional field.

Reviewer 2: ```suggestion
<13> Specify the file that has the `SSE-C` customer key to enable customer key encryption of the backups stored in S3. The file must contain a 32-byte string. The `customerKeyEncryptionFile` points to a mounted secret within the `velero` container. Add the following key-value pair to the `velero` `cloud-credentials` secret: `customer-key: <your_b64_encoded_32byte_string>`. Note that the `customerKeyEncryptionFile` field cannot be used with the `kmsKeyId`. The default value is an empty string (""), which means, `SSE-C` is disabled. This is an optional field.
```

> <13> Specify the file that has the `SSE-C` customer key to enable customer key encryption of the backups stored in S3. The file must contain a 32-byte string. The `customerKeyEncryptionFile` points to a mounted secret within the `velero` container. Add the following key-value pair to the `velero` `cloud-credentials` secret: `customer-key: <your_b64_encoded_32byte_string>`. Note that the `customerKeyEncryptionFile` field cannot be used with the `kmsKeyId`. The default value is an empty string (""), which means, `SSE-C` is disabled. This is an optional field.

Reviewer: These parameter names should have nouns after them for clarity:
```suggestion
<13> Specify the file that has the `SSE-C` customer key to enable customer key encryption of the backups stored in S3. The file must contain a 32-byte string. The `customerKeyEncryptionFile` field points to a mounted secret within the `velero` container. Add the following key-value pair to the `velero` `cloud-credentials` secret: `customer-key: <your_b64_encoded_32byte_string>`. Note that the `customerKeyEncryptionFile` field cannot be used with the `kmsKeyId` field. The default value is an empty string (""), which means, `SSE-C` is disabled. This is an optional field.
```

> <13> Specify the file that has the `SSE-C` customer key to enable customer key encryption of the backups stored in S3. The file must contain a 32-byte string. The `customerKeyEncryptionFile` points to a mounted secret within the `velero` container. Add the following key-value pair to the `velero` `cloud-credentials` secret: `customer-key: <your_b64_encoded_32byte_string>`. Note that the `customerKeyEncryptionFile` field cannot be used with the `kmsKeyId`. The default value is an empty string (""), which means, `SSE-C` is disabled. This is an optional field.

Reviewer: same point about backticks if it shows up in a file:
```suggestion
<13> Specify the file that has the `SSE-C` customer key to enable customer key encryption of the backups stored in S3. The file must contain a 32-byte string. The `customerKeyEncryptionFile` points to a mounted secret within the `velero` container. Add the following key-value pair to the `velero` `cloud-credentials` secret: `customer-key: <your_b64_encoded_32byte_string>`. Note that the `customerKeyEncryptionFile` field cannot be used with the `kmsKeyId`. The default value is an empty string (`""`), which means, `SSE-C` is disabled. This is an optional field.
```

Author: Done. thanks.

> <13> Specify the file that has the `SSE-C` customer key to enable customer key encryption of the backups stored in S3. The file must contain a 32-byte string. The `customerKeyEncryptionFile` points to a mounted secret within the `velero` container. Add the following key-value pair to the `velero` `cloud-credentials` secret: `customer-key: <your_b64_encoded_32byte_string>`. Note that the `customerKeyEncryptionFile` field cannot be used with the `kmsKeyId`. The default value is an empty string (""), which means, `SSE-C` is disabled. This is an optional field.

Reviewer: This comma can be removed as well:
```suggestion
<13> Specify the file that has the `SSE-C` customer key to enable customer key encryption of the backups stored in S3. The file must contain a 32-byte string. The `customerKeyEncryptionFile` points to a mounted secret within the `velero` container. Add the following key-value pair to the `velero` `cloud-credentials` secret: `customer-key: <your_b64_encoded_32byte_string>`. Note that the `customerKeyEncryptionFile` field cannot be used with the `kmsKeyId`. The default value is an empty string (""), which means `SSE-C` is disabled. This is an optional field.
```

Author: Done.

> <14> The version of the signature algorithm used to create signed URLs. You use signed URLs to download the backups, or fetch the logs. Possible versions are `1` and `4`. The default version is `4`. This is an optional field.

Reviewer: Typically we use the phrase valid values unless there's a particular reason to keep it this way:
```suggestion
<14> The version of the signature algorithm used to create signed URLs. You use signed URLs to download the backups, or fetch the logs. Valid values are `1` and `4`. The default version is `4`. This is an optional field.
```

Author: Done. thanks.

> <14> The version of the signature algorithm used to create signed URLs. You use signed URLs to download the backups, or to fetch the logs. Possible versions are `1` and `4`. The default version is `4`. This is an optional field.

Reviewer 2: ```suggestion
<14> The version of the signature algorithm used to create signed URLs. You use signed URLs to download the backups, or fetch the logs. Possible versions are `1` and `4`. The default version is `4`. This is an optional field.
```

> <15> The name of the {aws-short} profile in the credentials file. Default value is `default`. This is an optional field.

Reviewer: ```suggestion
<15> The name of the {aws-short} profile in the credentials file. The default value is `default`. This is an optional field.
```

Author: Done.

> <16> Set the `insecureSkipTLSVerify` field to `true` if you do not want to verify the TLS certificate when connecting to the object store, for example, for self-signed certificates with MinIO. Setting to `true` is susceptible to man-in-the-middle attacks and is not recommended for production workloads. Default value is `false`. This is an optional field.

Reviewer: ```suggestion
<16> Set the `insecureSkipTLSVerify` field to `true` if you do not want to verify the TLS certificate when connecting to the object store, for example, for self-signed certificates with MinIO. Setting to `true` is susceptible to man-in-the-middle attacks and is not recommended for production workloads. The default value is `false`. This is an optional field.
```

Author: Done.

> <17> Set the `enableSharedConfig` field to `true` if you want to load the credentials file as a shared config file. Default value is `false`. This is an optional field.
> <18> Specify the tags to annotate the {aws-short} S3 objects. Specify the tags in key-value pairs. Default value is an empty string (""). This is an optional field.
> <19> Specify the checksum algorithm to use for uploading objects to S3. The supported values are: `CRC32`, `CRC32C`, `SHA1`, and `SHA256`. If you set the field as an empty string (""), the checksum check will be skipped. Default value is `CRC32`. This is an optional field.

Reviewer: Just a few things that I've mentioned earlier in this review:
```suggestion
<17> Set the `enableSharedConfig` field to `true` if you want to load the credentials file as a shared config file. The default value is `false`. This is an optional field.
<18> Specify the tags to annotate the {aws-short} S3 objects. Specify the tags in key-value pairs. The default value is an empty string (`""`). This is an optional field.
<19> Specify the checksum algorithm to use for uploading objects to S3. The supported values are: `CRC32`, `CRC32C`, `SHA1`, and `SHA256`. If you set the field as an empty string (`""`), the checksum check will be skipped. The default value is `CRC32`. This is an optional field.
```

Author: Done.
